packages/sprotty-protocol/src/actions.ts

```diff
@@ -0,0 +1,500 @@
+    requestId: string
+}
+
+export function isRequestAction(object?: Action): object is RequestAction<ResponseAction> {
+    return hasOwnProperty<string, string>(object, 'requestId', 'string');
+}
+
+let nextRequestId = 1;
+/**
+ * Generate a unique `requestId` for a request action.
+ */
+export function generateRequestId(): string {
+    return (nextRequestId++).toString();
+}
+
+/**
+ * A response action is sent to respond to a request action. The `responseId` must match
+ * the `requestId` of the preceding request. In case the `responseId` is empty or undefined,
+ * the action is handled as standalone, i.e. it was fired without a preceding request.
+ */
+export interface ResponseAction extends Action {
+    responseId: string;
+}
+
+/**
+ * A response action is sent to respond to a request action. The `responseId` must match
+ * the `requestId` of the preceding request. In case the `responseId` is empty or undefined,
+ * the action is handled as standalone, i.e. it was fired without a preceding request.
+ */
+
+export function isResponseAction(object?: Action): object is ResponseAction {
+    return hasOwnProperty<string, string>(object, 'responseId', 'string');
+}
+
+/**
+ * A reject action is fired to indicate that a request must be rejected.
+ */
+export interface RejectAction extends ResponseAction {
+    kind: typeof RejectAction.KIND
+    message: string
+    detail?: JsonAny
+}
+export namespace RejectAction {
+    export const KIND = 'rejectRequest';
+
+    export function create(options: { message: string, detail?: JsonAny, requestId: string }): RejectAction {
+        return {
+            kind: KIND,
+            message: options.message,
+            detail: options.detail,
+            responseId: options.requestId
+        };
+    }
+}
+
+/**
+ * Sent from the client to the model source (e.g. a DiagramServer) in order to request a model. Usually this
+ * is the first message that is sent to the source, so it is also used to initiate the communication.
+ * The response is a SetModelAction or an UpdateModelAction.
+ */
+export interface RequestModelAction extends RequestAction<SetModelAction> {
+    kind: typeof RequestModelAction.KIND
+    options?: { [key: string]: string | number | boolean }
+}
+export namespace RequestModelAction {
+    export const KIND = 'requestModel';
+
+    export function create(options?: { [key: string]: JsonPrimitive }): RequestModelAction {
+        return {
+            kind: KIND,
+            options,
+            requestId: generateRequestId()
+        };
+    }
+}
+
+/**
+ * Sent from the model source to the client in order to set the model. If a model is already present, it is replaced.
+ */
+export interface SetModelAction extends ResponseAction {
+    kind: typeof SetModelAction.KIND
+    newRoot: SModelRoot
+}
+export namespace SetModelAction {
+    export const KIND = 'setModel';
+
+    export function create(newRoot: SModelRoot, requestId?: string): SetModelAction {
```

Comment:
The create function for response actions should be initialized with an empty request id (to enable standalone response actions)
```suggestion
    export function create(newRoot: SModelRoot, requestId=''): SetModelAction {
```